src/main/java/org/openrewrite/java/testing/cleanup/RemoveVisibleForTestingAnnotationWhenUsedInProduction.java

```diff
@@ -0,0 +1,176 @@
+                "This recipe should not be used in an environment where QA tooling acts on the `@VisibleForTesting` annotation.";
+    }
+
+    @Override
+    public Scanned getInitialValue(ExecutionContext ctx) {
+        return new Scanned();
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getScanner(Scanned acc) {
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            @Override
+            public J.FieldAccess visitFieldAccess(J.FieldAccess fieldAccess, ExecutionContext ctx) {
+                // Mark classes
+                if (fieldAccess.getTarget().getType() instanceof JavaType.Class) {
+                    checkAndRegister(acc.classes, fieldAccess.getTarget().getType());
+                }
+                // Mark fields
+                if (fieldAccess.getName().getFieldType() != null) {
+                    checkAndRegister(acc.fields, fieldAccess.getName().getFieldType());
+                }
+                return super.visitFieldAccess(fieldAccess, ctx);
+            }
+
+            @Override
+            public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+                if (method.getMethodType() != null) {
+                    checkAndRegister(acc.methods, method.getMethodType());
+                }
+                return super.visitMethodInvocation(method, ctx);
+            }
+
+            @Override
+            public J.NewClass visitNewClass(J.NewClass newClass, ExecutionContext ctx) {
+                // Mark constructors
+                if (newClass.getConstructorType() != null) {
+                    checkAndRegister(acc.methods, newClass.getConstructorType());
+                }
+                // Mark classes
+                if (newClass.getClazz() != null && newClass.getClazz().getType() != null && newClass.getClazz().getType() instanceof JavaType.Class) {
+                    checkAndRegister(acc.classes, newClass.getClazz().getType());
+                }
+                return super.visitNewClass(newClass, ctx);
+            }
+
+            private void checkAndRegister(List<String> target, JavaType type) {
+                if (!target.contains(TypeUtils.toString(type))) {
+                    getAnnotations(type).forEach(annotation -> {
+                        if ("VisibleForTesting".equals(annotation.getClassName())) {
+                            J.CompilationUnit compilationUnit = getCursor().firstEnclosing(J.CompilationUnit.class);
+                            if (compilationUnit != null) {
+                                compilationUnit
+                                        .getMarkers()
+                                        .findFirst(JavaSourceSet.class)
+                                        .filter(elem -> "main".equals(elem.getName()))
+                                        .ifPresent(sourceSet -> target.add(TypeUtils.toString(type)));
+                            }
```

Comment:
I think I would write it more like:

```suggestion
                            if (cu != null &&
                                   cu.getMarkers().findFirst(JavaSourceSet.class).filter(s -> s.getName().toLowerCase(Locale.ROOT).contains("main")).isPresent()) {
                                   target.add(TypeUtils.toString(type))
                            }
```

Reply:
Can you elaborate on this change?
s -> s.getName().toLowerCase(Locale.ROOT).contains("main")

It seems there are several considerations when trying to assert a file is in the production code?

Reply:
I don't know, but I always try to find an existing example. Something you did here is also used in:

https://github.com/openrewrite/rewrite/blob/c08ea6273755822643bdb6d95e8350b50abf5be3/rewrite-java/src/main/java/org/openrewrite/java/search/IsLikelyTest.java#L65